src/tainting/Dataflow_tainting.ml

```diff
@@ -435,48 +435,224 @@ let partition_mutating_sources sources_matches =
          match t.orig with
          | Src src -> Some src.label
          | Arg _ -> None)
-  |> LabelSet.of_seq
+  |> LabelSet.of_list
 
 (* coupling: if you modify the code here, you may need to modify 'Parse_rule.parse_taint_requires' too. *)
 let rec eval_label_requires ~labels e =
-  match e.G.e with
-  | G.L (G.Bool (v, _)) -> v
-  | G.N (G.Id (id, _)) ->
-      let str, _ = id in
-      LabelSet.mem str labels
-  | G.Call ({ e = G.IdSpecial (G.Op G.Not, _); _ }, (_, [ Arg e1 ], _)) ->
-      not (eval_label_requires ~labels e1)
-  | G.Call ({ e = G.IdSpecial (G.Op op, _); _ }, (_, args, _)) -> (
-      match (op, eval_label_args ~labels args) with
-      | G.And, xs -> List.fold_left ( && ) true xs
-      | G.Or, xs -> List.fold_left ( || ) false xs
-      | __else__ ->
-          logger#error "Unexpected Boolean operator";
-          false)
-  | G.ParenExpr (_, e, _) -> eval_label_requires ~labels e
-  | ___else__ ->
-      logger#error "Unexpected `requires' expression";
-      false
-
-and eval_label_args ~labels args =
-  match args with
-  | [] -> []
-  | G.Arg e :: args' ->
-      eval_label_requires ~labels e :: eval_label_args ~labels args'
-  | _ :: args' ->
-      logger#error "Unexpected argument kind";
-      false :: eval_label_args ~labels args'
-
-let taints_satisfy_requires taints expr =
-  let labels = labels_of_taints taints in
+  match e with
+  | T.Bool v -> v
+  | Label str -> LabelSet.mem str labels
+  | Not e -> not (eval_label_requires ~labels e)
+  | And xs ->
+      xs
+      |> Common.map (eval_label_requires ~labels)
+      |> List.fold_left ( && ) true
+  | Or xs ->
+      xs
+      |> Common.map (eval_label_requires ~labels)
+      |> List.fold_left ( || ) false
+
+(* This is where we actually do the solving of the "hypothetical taint".
+   To recap, to deal with taint which may or may not exist due to the
+   later value of taint variables, we need to attach a "precondition" to
+   such hypothetical taints, detailing the conditions upon which they can
+   appear.
+
+   We do this super lazily, because we can't know whether it's solvable until
+   we actually substitute for any argument taint variables that may be involved.
+   Once that substitution has been done, however, we need to solve to figure out
+   if this hypothetical taint could exist after all.
+
+   This can only happen if the precondition formula is satisfied by all of
+   the dependency taints that really do "exist", in the sense that they are known
+   to be un-hypothetical and not polymorphic, or hypothetical and their preconditions
+   are solved in the same way.
+
+   That's what this does.
+*)
+let rec taint_can_exist taint =
+  match taint.T.orig with
+  | Arg _ -> false
+  | Src { precondition = None; _ } -> true
+  | Src { precondition = Some (incoming, expr); _ } ->
+      solve_precondition (incoming, expr)
+
+and solve_precondition (incoming, expr) =
+  let labels = List.filter taint_can_exist incoming |> labels_of_taints in
```

Comment:
So we just ignore the `Arg` taints when evaluating the requires?

Reply:
Yep, under hypothesis that any such `Arg` taints we encounter should have been substituted for, already. This isn't always the case, because we might run `pm_of_finding` on a finding within a function call, but then ignoring them is
the right thing to do, because we don't yet have enough information on the `Arg` taint (that will be instantiated in a different finding, if the function is called at all)